### ISSUE TITLE
Uncaught exception 'Drupal\Core\Database\ConnectionNotDefinedException'

### ISSUE THREAD

**Issue by valkum** [drush-ops/drush#1232]
I've got a problem. Somehow the database settings set in my site-alias are not set for drupal.

This is my error:
```
valkum@server2:/var/www/news.page.com/sites/default$ drush @prod status
PHP Fatal error:  Uncaught exception 'Drupal\Core\Database\ConnectionNotDefinedException' with message 'The specified database connection is not defined: default' in /var/www/news.page.com/core/lib/Drupal/Core/Database/Database.php:366
Stack trace:
#0 /var/www/news.page.com/core/lib/Drupal/Core/Database/Database.php(171): Drupal\Core\Database\Database::openConnection('default', 'default')
#1 /var/www/news.page.com/sites/default/files/php/service_container/service_container_prod/d22d714c78c5d5a7f279061e951e6e29625f03fbe25769e8e8360cd0f1da25a4.php(2345): Drupal\Core\Database\Database::getConnection('default')
#2 /var/www/news.page.com/core/vendor/symfony/dependency-injection/Symfony/Component/DependencyInjection/Container.php(326): Drupal\Core\DependencyInjection\Container\prod\service_container_prod->getDatabaseService()
#3 /var/www/news.page.com/sites/default/files/php/service_container/service_container_prod/d22d714c78c5d5a7f279061e951e6e29625f03fbe25769e8e8360cd0f1da25a4.php( in /var/www/news.page.com/core/lib/Drupal/Core/Database/Database.php on line 366
PHP Fatal error:  Uncaught exception 'Drupal\Core\Database\ConnectionNotDefinedException' with message 'The specified database connection is not defined: default' in /var/www/news.page.com/core/lib/Drupal/Core/Database/Database.php:366
Stack trace:
#0 /var/www/news.page.com/core/lib/Drupal/Core/Database/Database.php(171): Drupal\Core\Database\Database::openConnection('default', 'default')
#1 /var/www/news.page.com/sites/default/files/php/service_container/service_container_prod/d22d714c78c5d5a7f279061e951e6e29625f03fbe25769e8e8360cd0f1da25a4.php(2345): Drupal\Core\Database\Database::getConnection('default')
#2 /var/www/news.page.com/core/vendor/symfony/dependency-injection/Symfony/Component/DependencyInjection/Container.php(326): Drupal\Core\DependencyInjection\Container\prod\service_container_prod->getDatabaseService()
#3 /var/www/news.page.com/sites/default/files/php/service_container/service_container_prod/d22d714c78c5d5a7f279061e951e6e29625f03fbe25769e8e8360cd0f1da25a4.php( in /var/www/news.page.com/core/lib/Drupal/Core/Database/Database.php on line 366

```
In my settings.php for that site i have:
```
$db_creds = unserialize(getenv('DRUPAL_DB'));
if($db_creds != false) {
  $databases = array (
    'default' => array (
      'default' => $db_creds,
    ),
  );
  $databases['default']['default']['namespace'] = 'Drupal\\Core\\Database\\Driver\\mysql';
}
```
DRUPAL_DB is set in my vhost file via SetEnv and contains an Database array with my credentials.
My site-alias file:
```
$aliases['prod'] = array(
  'uri' => 'news.page.com',
  'root' => '/var/www/news.page.com',
  'os' => 'Linux',
  'db-url' => 'mysql://news_page:defaced@localhost/news_page'
  'databases' =>
  array (
    'default' =>
    array (
      'driver' => 'mysql',
      'username' => 'news_page',
      'password' => 'defaced',
      'port' => '',
      'host' => 'localhost',
      'database' => 'news_page',
      'namespace' => 'Drupal\\Core\\Database\\Driver\\mysql',
    ),
  ),
);
```
Seems strange. Normally the $databases variable should be overwritten with the one from the file alias, or am i wrong?

**Comment by weitzman** [comment]
The alias db-url is not used for all commands. It is mainly only used for sql-* commands. Otherwise, Drupal bootstraps itself and the error you get happens because virtual hosts are not in play for CLI requests. 

Fortunately, you can pass ENV variables in an alias. See #955. Requires Drush7. Once you pass that, I think you can remove db-url and databases from the alias if you want.

**Comment by valkum** [comment]
Okay thanks for this information. But i think #955 doesn't fix this in a stable way.
This is my alias now. When i add an die(phpinfo()); for debugging to my settings.php there is no Env Variable set for DRUPAL_DB.
´´´
$aliases['prod'] = array(
  'uri' => 'news.globalgameport.com',
  'root' => '/var/www/news.globalgameport.com',
  'site_path' => '/var/www/news.globalgameport.com/sites/default',
  '#env-vars' => array(
    'DRUPAL_DB' => 'a:6:{s:6:"driver";s:5:"mysql";s:8:"database";s:9:"news_page";s:8:"username";s:9:"news_page";s:8:"password";s:16:"defaced";s:4:"host";s:9:"localhost";s:6:"prefix";s:0:"";}'
  ),
);
´´´

I tracked the possible fix of #955 down to drush_invoke_process. But this function is not used always. Calls with drush_invoke doesn't set any prefix with env vars.

As there is no possible solution for this i have to go back to write my config into settings.php.
